op-txproxy/auth_handler.go

```diff
@@ -0,0 +1,95 @@
+package op_txproxy
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var (
+	defaultBodyLimit = 5 * 1024 * 1024 // default in op-geth
+
+	DefaultAuthHeaderKey = "x-optimism-signature"
+)
+
+// authHandler detects when authentication information is present on the request.
+// If so, it will validate and set authenicated caller in the request context. It does not
+// reject if authentication information is missing. It is up to the request handler to do so.
+type authHandler struct {
+	headerKey string
+	next      http.Handler
+}
+
+func AuthMiddleware(headerKey string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return &authHandler{headerKey, next}
+	}
+}
+
+type authContextKey struct{}
+
+type AuthContext struct {
+	Caller common.Address
+}
+
+// ServeHTTP serves JSON-RPC requests over HTTP, implements http.Handler
+func (h *authHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	authHeader := r.Header.Get(h.headerKey)
+	if authHeader == "" {
+		h.next.ServeHTTP(w, r)
+		return
+	}
+	authElems := strings.Split(authHeader, ":")
+	if len(authElems) != 2 {
+		http.Error(w, "misformatted auth header", http.StatusBadRequest)
+		return
+	}
+
+	if r.Body == nil {
+		// edge case from unit tests
+		r.Body = io.NopCloser(bytes.NewBuffer(nil))
+	}
+
+	// The request body can only be read once. Adjust here so it's inspectable
+	var buf bytes.Buffer
+	body := io.TeeReader(r.Body, &buf)
+	r.Body = io.NopCloser(&buf)
+
+	// Since this middleware runs prior to the server, we need to manually apply the body limit when reading.
+	bodyBytes, err := io.ReadAll(io.LimitReader(body, int64(defaultBodyLimit)))
+	if err != nil {
+		http.Error(w, "unable to parse request body", http.StatusInternalServerError)
+		return
+	}
+
+	txtHash := accounts.TextHash(bodyBytes)
+	caller, signature := common.HexToAddress(authElems[0]), common.FromHex(authElems[1])
+	sigPubKey, err := crypto.SigToPub(txtHash, signature)
+	if err != nil {
+		http.Error(w, "invalid authentication signature", http.StatusBadRequest)
+		return
+	}
+
+	if caller != crypto.PubkeyToAddress(*sigPubKey) {
+		http.Error(w, "mismatched recovered signer", http.StatusBadRequest)
+		return
+	}
+
+	// Set the authenticated caller in the context
+	newCtx := context.WithValue(r.Context(), authContextKey{}, &AuthContext{caller})
+	h.next.ServeHTTP(w, r.Clone(newCtx))
```

Comment:
The Request.Clone is good, but I think it's too conservative here. I think WithContext is fine because you haven't modified the original request (except the Body and that's fine). I'm curious what you think here.

Reply:
Ahh I missed `WithContext` on request. Yeah that's much better